### PR TITLE
Listen thread

### PR DESCRIPTION
**Pull request by logankuzyk**

### CODE REVIEW

File: apps/VideoStream/videoStream.py [hunk 1]
```diff
@@ -68,10 +69,16 @@ def send_frame(self, frame, address, quality=4):
...
+    def listen_thread(self):
+        thread = threading.Thread(target = self.listen)
+        thread.daemon = False
+        thread.start()
+
     def listen(self, port=None):
         if port is None:
             port = self.port
-        self.socket.bind(('', port))
+        with threading.Lock():
```
Comment (dragonprevost) [inline apps/VideoStream/videoStream.py]: Might be nice to put a short sleep cycle in this loop, maybe a second or 2 so that it does not take up too many resources on the pi.

**Comment by logankuzyk** [comment]
I implemented a new technique to listen for new connections which prevents race conditions and thread blocking.

There was an error in the `connections.add()` method, which set the initial connection time as 0 instead of `time.time()`. This caused all connections to be removed every time `connections.cleanup()` got called, because the expression for the delta would simplify to `time.time() - 0`. There were also issues with disconnecting and reconnecting to the video server due to my incorrect usage of locks.

Main and listen are still the only two threads in the program. The `listen()` method got changed the most. Instead of being constantly blocked by `socket.recvfrom()` and sleeping, I used `socket.timeout` and an exception handler to prevent the thread from blocking `broadcast()` when updating connections.

With this new design, `listen()` can continuously loop with only one thread. This allows for the periodic execution of `connections.cleanup()` and avoiding down time while listening by not using sleep (connections would be rejected if they tried to connect when listen was sleeping).

**Comment by dragonprevost** [comment]
> nnections.cleanup

Good eye finding that bug, those things can be difficult to diagnose! Next meeting I am just going to get you to go over how you are testing the module, I will get it running on my machine then we can get everything merged up!

File: apps/VideoStream/videoStream.py [hunk 2]
```diff
@@ -43,50 +41,63 @@ def update(self, address):
...
+                data = data.decode('utf-8')
+                if (data == "get"):
+                    self.connections.lock.acquire()
+                    self.connections.update(address)
+                    for address in self.connections.connections:
+                        print(address)
+                    self.connections.lock.release()
+                continue
+            except socket.timeout:
+                continue
```
[the remark this answers is not in the document]
Comment (logankuzyk) [reply apps/VideoStream/videoStream.py]: The continue statement is needed here because the socket.timeout exception is meant to happen periodically. In the case where videostream has been turned on and is waiting for new connections, the socket.timeout exception gets thrown after one second of waiting.

I can explain better in person, but I basically replaced sleeping with socket.timeout. I did this because the timeout prevents the loop from taking up too much resources while still allowing new connections to be made at any time.

To summarize, socket.recvfrom() is blocking until the bytes are actually received, but there are things that the listen thread needs to do periodically, such as cleaning up the connections and printing current connections. My solution to avoid making another thread was to use the socket.timeout property to re-run the loop at an interval, which is why the loop continues after the socket.timeout exception.

[the remark this answers is not in the document]
Comment (avesk) [reply apps/VideoStream/videoStream.py]: Gotcha

File: apps/VideoStream/videoStream.py [hunk 3]
```diff
@@ -8,36 +8,32 @@ class Connections:
...
-        self.times_since_heartbeat[index] = time.time()
 
     def cleanup(self):
-        i = 0
-        while i < len(self.times_since_heartbeat):
-            if(time.time() - self.times_since_heartbeat[i] > self.timeout):
-                print("Client", self.connections[i], "timed out.")
-                self.remove(i)
-            i = i + 1
+        to_remove = []
```
Comment (logankuzyk) [inline apps/VideoStream/videoStream.py]: Had to make a list of the connections to be removed. Removing them during the for loop caused a runtime error due to the dictionary changing in size.
